mPower2/mPower2/PassiveCollectors.swift

```diff
@@ -225,25 +235,456 @@ class PassiveDisplacementCollector : NSObject, PassiveCollector, CLLocationManag
+                self.stopRecorderAndUpload()
+            }
+
+            self.backgroundTaskId = UIApplication.shared.beginBackgroundTask(withName: "Passive gait collection") {
+                // if we get a callback that we're out of time before our timer expires, upload whatever we managed to get
+                #if DEBUG
+                let elapsedTime = Date().timeIntervalSince(self.startTime!)
+                self.debugNotification(title: "Cutting it short", body: "background task about to expire after \(elapsedTime) seconds")
+                #endif
+                self.stopRecorderAndUpload()
+            }
+        }
+    }
+    
+    private func endBackgroundTask() {
+        UIApplication.shared.endBackgroundTask(self.backgroundTaskId)
+        self.backgroundTaskId = .invalid
+    }
+    
+    #if DEBUG
+    func archiveErrorDesc(_ error: Error?) -> String {
+        guard let error = error else { return "No errors reported" }
+        guard let nserror = error as NSError?,
+                let underlyingError = nserror.userInfo[NSUnderlyingErrorKey] as? NSError
+            else {
+                return error.localizedDescription
+        }
+        let shortDomain = nserror.domain.components(separatedBy: ".").last ?? "???"
+        let ulShortDomain = underlyingError.domain.components(separatedBy: ".").last ?? "???"
+        return "Err cd \(nserror.code) - \(shortDomain): UL Err Cd \(underlyingError.code) - \(ulShortDomain) \(underlyingError.localizedDescription)"
+    }
+    #endif
+    
+    func stopRecorderAndUpload() {
+        guard let taskState = self.viewModel,
+                let recorder = self.recorder,
+                let timer = self.timer,
+                let startTime = self.startTime
+            else {
+                return
+        }
+        self.timer = nil
+        self.viewModel = nil
+        self.startTime = nil
+        self.recorder = nil
+
+        // kill the timer if it's still running
+        if timer.isValid {
+            timer.invalidate()
+        }
+        
+        #if DEBUG
+        debugNotification(title: "Stopping motion sensors", body: "as of \(Date())")
+        #endif
+        
+        recorder.stop() { (action, result, error) in
+            let duration = Date().timeIntervalSince(startTime)
+            
+            if error == nil, let result = result, duration >= 15.0 {
+                #if DEBUG
+                self.debugNotification(title: "Archiving", body: "\(taskState)")
+                #endif
+                
+                taskState.taskResult.appendAsyncResult(with: result)
+                
+                self.archiveManager.archiveAndUpload(taskState) { (_ error: Error?) in
+                    #if DEBUG
+                    self.debugNotification(title: "Done archiving", body: self.archiveErrorDesc(error))
+                    #endif
+                    self.endBackgroundTask()
```

Comment:
Why do you end the background task *after* archiving and uploading?

Reply:
because I was under the mistaken impression that the completion handler would be called once the upload request had been fired off in the background session, forgetting that in fact I built the upload manager such that the completion handler doesn't get called until the upload is fully done and confirmed. derp